### PR TITLE
Implement a minimal API for CPython

### PR DESCRIPTION
**Pull request by brettcannon**
API in CPython becomes:

1. `JitInit()`
2. `EvalFrame()`
3. Code objects gain `co_extra`
4. PyEval_EvalFrameEx_NoJit() (which should be named something else eventually)
5. Interpreter state gains `eval_frame`

Everything else is contained within Pyjion itself.

### CODE REVIEW

**Comment by aisk** [comment]
I think it's better to add a namespace prefix to `JitInit` functions like `Pyjion_JitInit`, because some one may be embed python in there C/C++ project, and `JitInit` may be have conflict with other codes.

**Comment by brettcannon** [comment]
From a nice, clean implementation you would be right, @aisk , but since this is currently just  a proof-of-concept I'm not too worried about name clashes.

**Comment by brettcannon** [comment]
One enhancement I may do if this is accepted is instead of having PyEval_EvalFrameEx() do an `if` check for `interp->eval_frame`, we could make it always do `interp->eval_frame` and by default set it to `PyEval_EvalFrameEx_NoJit()`. This would allow JITs to potentially stack themselves (something Numba may want).

File: Pyjion/pyjit.cpp [hunk 1]
```diff
@@ -75,20 +108,43 @@ extern "C" __declspec(dllexport) PyJittedCode* JitCompile(PyCodeObject* code) {
...
+static PY_UINT64_T HOT_CODE = 20000;
+
+extern "C" __declspec(dllexport) PyObject *EvalFrame(PyFrameObject *f, int throwflag) {
+    if (f->f_code->co_extra == nullptr) {
+        auto jitted = jittedcode_new_direct();
+        f->f_code->co_extra = (PyObject *)jitted;
+        jitted->j_run_count++;
+    }
+    else if (!throwflag) {
+        auto jitted = (PyjionJittedCode *)f->f_code->co_extra;
```
Comment (DinoV) [inline Pyjion/pyjit.cpp]: Should have a type check here to make sure it's a PyjionJittedCode object?

File: Pyjion/pyjit.cpp [hunk 2]
```diff
@@ -75,20 +108,43 @@ extern "C" __declspec(dllexport) PyJittedCode* JitCompile(PyCodeObject* code) {
...
+    g_pyjionJittedCode[jittedCode] = res;
     jittedCode->j_evalfunc = &Jit_EvalHelper;
     jittedCode->j_evalstate = res->get_code_addr();
-    return jittedCode;
+    return true;
+}
+
+static PY_UINT64_T HOT_CODE = 20000;
+
+extern "C" __declspec(dllexport) PyObject *EvalFrame(PyFrameObject *f, int throwflag) {
```
Comment (DinoV) [inline Pyjion/pyjit.cpp]: I'm a little curious if we want to flow throwflag through here or not... It seems like this is basically an internal Python implementation and we just never want to pass the flag through (e.g. not call the external handler when it's set, and then assume it's always 0 when executing).

File: Pyjion/pyjit.cpp [hunk 3]
```diff
@@ -75,20 +108,43 @@ extern "C" __declspec(dllexport) PyJittedCode* JitCompile(PyCodeObject* code) {
...
+    if (f->f_code->co_extra == nullptr) {
+        auto jitted = jittedcode_new_direct();
+        f->f_code->co_extra = (PyObject *)jitted;
```
Comment (DinoV) [inline Pyjion/pyjit.cpp]: Null check for OOM?

File: Pyjion/pyjit.cpp (hunk 2 above)
Comment (brettcannon) [inline Pyjion/pyjit.cpp]: As discussed elsewhere privately, we do if we want to follow my idea of simply assigning `PyEval_EvalFrameEx()` to the slot and generalizing this even farther.

**Comment by brettcannon** [comment]
PTAL; addressed your comments, fixed some code that was broken after the merge w/ the tracing work, and added a batch file to make it easier to re-generate the patchset for Python.

File: Pyjion/pyjit.cpp [hunk 4]
```diff
@@ -335,6 +303,42 @@ __declspec(dllexport) bool jit_compile(PyCodeObject* code) {
...
+            return NULL;
+        }
+
+        f->f_code->co_extra = (PyObject *)jitted;
+        jitted->j_run_count++;
+    }
+    else if (!throwflag) {
+        auto jitted = (PyjionJittedCode *)f->f_code->co_extra;
+        if (Py_TYPE(jitted) != &PyjionJittedCode_Type) {
+            return NULL;
```
Comment (DinoV) [inline Pyjion/pyjit.cpp]: Shouldn't this do PyEval_EvalFrameEx_NoJit?

**Comment by brettcannon** [comment]
PTAL; simplified the `if` statement

**Comment by brettcannon** [comment]
Dino cleared this PR offline

**Pull request merged**